File: src/grok/mcpSpec.ts
import { existsSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import type { DeployConfig, ReadPolicy } from "../deployConfig.js";
import { conversationTypeFromSurface, surfaceEnvVars, type Surface } from "../surface.js";
import {
  MCP_TEAMS_READ_SERVER,
  MCP_TEAMS_WRITE_SERVER,
  serverNameForRole,
  writeScopeFromSurface,
  type McpRole,
  type WriteScope,
} from "../mcp/tools.js";
import type { AcpEnvVar, AcpMcpServerStdio, TurnId } from "../types.js";

function mcpServerEntry(): { command: string; args: string[] } {
  const here = dirname(fileURLToPath(import.meta.url));
  const js = join(here, "..", "mcpServerMain.js");
  const ts = join(here, "..", "mcpServerMain.ts");
  if (existsSync(js)) {
    return { command: process.execPath, args: [js, "--stdio"] };
  }
  const tsxBin = join(process.cwd(), "node_modules", ".bin", "tsx");
  const tsx = existsSync(tsxBin) ? tsxBin : "tsx";
  return { command: process.execPath, args: [tsx, ts, "--stdio"] };
}

function inheritedEnv(): AcpEnvVar[] {
  const names = ["PATH", "HOME", "USER", "TMPDIR", "LANG", "NODE_PATH", "GRAPH_TOKEN", "GRAPH_CLIENT_ID", "GRAPH_CLIENT_SECRET", "GRAPH_TENANT_ID"];
  const out: AcpEnvVar[] = [];
  for (const name of names) {
    const value = process.env[name];
    if (value !== undefined && value.length > 0) {
      out.push({ name, value });
    }
  }
  return out;
}

function readPolicyEnv(policy: ReadPolicy, readRecentThreads: boolean): AcpEnvVar[] {
  const env: AcpEnvVar[] = [
    { name: "READ_POLICY", value: policy.kind },
    { name: "FEATURES_READ_RECENT_THREADS", value: readRecentThreads ? "true" : "false" },
  ];
  if (policy.kind === "allowlist") {
    env.push({ name: "READ_CHANNELS", value: JSON.stringify(policy.channels) });
  }
  return env;
}

function commonEnv(args: {
  turnId: TurnId;
  turnsDir: string;
  callbackUrl?: string;
  conversationId: string;
  serviceUrl: string;
  conversationType: string;
  surface?: Surface;
  readPolicy?: ReadPolicy;
  readRecentThreads?: boolean;
}): AcpEnvVar[] {
  const env: AcpEnvVar[] = [
    ...inheritedEnv(),
    { name: "TURN_ID", value: args.turnId },
    { name: "TURNS_DIR", value: args.turnsDir },
    { name: "CONVERSATION_ID", value: args.conversationId },
    { name: "SERVICE_URL", value: args.serviceUrl },
    {
      name: "CONVERSATION_TYPE",
      value: args.surface !== undefined ? conversationTypeFromSurface(args.surface) : args.conversationType,
    },
  ];
  if (args.callbackUrl !== undefined && args.callbackUrl.length > 0) {
    env.push({ name: "HOST_CALLBACK_URL", value: args.callbackUrl });
  }
  if (args.surface !== undefined) {
    env.push(...surfaceEnvVars(args.surface));
  }
  if (args.readPolicy !== undefined) {
    env.push(...readPolicyEnv(args.readPolicy, args.readRecentThreads === true));
  }
  return env;
}

function writeScopeEnv(scope: WriteScope, conversationId: string): AcpEnvVar[] {
  switch (scope.kind) {
    case "chat":
      return [
        { name: "WRITE_KIND", value: "chat" },
        { name: "WRITE_CONVERSATION_ID", value: scope.conversationId },
      ];
    case "channel":
      return [
        { name: "WRITE_KIND", value: "channel" },
        { name: "WRITE_CONVERSATION_ID", value: conversationId },
        { name: "WRITE_TEAM_ID", value: scope.teamId },
        { name: "WRITE_CHANNEL_ID", value: scope.channelId },
        { name: "WRITE_THREAD_ID", value: scope.threadId },
      ];
    default: {
      const _exhaustive: never = scope;
      throw new Error(`unknown write scope ${JSON.stringify(_exhaustive)}`);
    }
  }
}

function stdioSpec(args: {
  role: McpRole;
  env: AcpEnvVar[];
}): AcpMcpServerStdio {
  const launched = mcpServerEntry();
  return {
    type: "stdio",
    name: serverNameForRole(args.role),
    command: launched.command,
    args: [...launched.args, `--role=${args.role}`],
    env: args.env,
  };
}

export function teamsMcpServers(args: {
  turnId: TurnId;
  turnsDir: string;
  callbackUrl?: string;
  conversationId: string;
  serviceUrl: string;
  conversationType: string;
  writeScope?: WriteScope;
  surface?: Surface;
  deploy?: DeployConfig;
  readPolicy?: ReadPolicy;
  readRecentThreads?: boolean;
}): { read: AcpMcpServerStdio; write: AcpMcpServerStdio } {
  const surface = args.surface;
  const writeScope = args.writeScope ?? (surface !== undefined ? writeScopeFromSurface(surface) : undefined);
  if (writeScope === undefined) {
    throw new Error("teamsMcpServers requires surface or writeScope");
  }
  const readPolicy = args.readPolicy ?? args.deploy?.readPolicy;
  const readRecentThreads = args.readRecentThreads ?? args.deploy?.features.readRecentThreads ?? false;
  const shared = commonEnv({
    turnId: args.turnId,
    turnsDir: args.turnsDir,
    conversationId: args.conversationId,
    serviceUrl: args.serviceUrl,
    conversationType: args.conversationType,
    ...(args.callbackUrl !== undefined ? { callbackUrl: args.callbackUrl } : {}),
    ...(surface !== undefined ? { surface } : {}),
    ...(readPolicy !== undefined ? { readPolicy, readRecentThreads } : {}),
  });
  const read = stdioSpec({ role: "read", env: shared });
  const write = stdioSpec({
    role: "write",
    env: [...shared, ...writeScopeEnv(writeScope, args.conversationId)],
  });
  return { read, write };
}

/** Grok config.json/GROK_CONFIG form. Omit env so the MCP child inherits grok's process env. */
export function grokConfigForMcp(read: AcpMcpServerStdio, write: AcpMcpServerStdio): string {
  return JSON.stringify({
    mcp_servers: {
      [MCP_TEAMS_READ_SERVER]: {
        command: read.command,
        args: read.args,
      },
      [MCP_TEAMS_WRITE_SERVER]: {
        command: write.command,
        args: write.args,
      },
    },
  });
}

File: python/teams_harness/mcp_spec.py
from __future__ import annotations

import os
import shutil
from pathlib import Path

from .tools import MCP_TEAMS_READ_SERVER, MCP_TEAMS_WRITE_SERVER, McpRole, server_name_for_role
from .types import AcpEnvVar, AcpMcpServerStdio, TurnId


def repo_root() -> Path:
    here = Path(__file__).resolve()
    for parent in [here.parent, *here.parents]:
        if (parent / "src" / "mcpServerMain.ts").exists() or (parent / "package.json").exists():
            if (parent / "src").is_dir():
                return parent
    return Path.cwd()


def mcp_server_entry() -> tuple[str, list[str]]:
    root = repo_root()
    js = root / "src" / "mcpServerMain.js"
    ts = root / "src" / "mcpServerMain.ts"
    # Prefer the Node MCP server that already exists in this repo. Python never talks Graph.
    node_bin = shutil.which("node") or "node"
    if js.exists():
        return node_bin, [str(js), "--stdio"]
    tsx_bin = root / "node_modules" / ".bin" / "tsx"
    tsx = str(tsx_bin) if tsx_bin.exists() else "tsx"
    return node_bin, [tsx, str(ts), "--stdio"]


def inherited_env() -> list[AcpEnvVar]:
    names = [
        "PATH",
        "HOME",
        "USER",
        "TMPDIR",
        "LANG",
        "NODE_PATH",
        "GRAPH_TOKEN",
        "GRAPH_CLIENT_ID",
        "GRAPH_CLIENT_SECRET",
        "GRAPH_TENANT_ID",
    ]
    out: list[AcpEnvVar] = []
    for name in names:
        value = os.environ.get(name)
        if value:
            out.append({"name": name, "value": value})
    return out


def common_env(
    *,
    turn_id: TurnId,
    turns_dir: str,
    conversation_id: str,
    service_url: str,
    conversation_type: str,
    callback_url: str | None = None,
) -> list[AcpEnvVar]:
    env: list[AcpEnvVar] = [
        *inherited_env(),
        {"name": "TURN_ID", "value": turn_id},
        {"name": "TURNS_DIR", "value": turns_dir},
        {"name": "CONVERSATION_ID", "value": conversation_id},
        {"name": "SERVICE_URL", "value": service_url},
        {"name": "CONVERSATION_TYPE", "value": conversation_type},
    ]
    if callback_url:
        env.append({"name": "HOST_CALLBACK_URL", "value": callback_url})
    return env


def write_scope_env(scope: dict[str, str], conversation_id: str) -> list[AcpEnvVar]:
    kind = scope.get("kind")
    if kind == "chat":
        return [
            {"name": "WRITE_KIND", "value": "chat"},
            {"name": "WRITE_CONVERSATION_ID", "value": scope["conversationId"]},
        ]
    if kind == "channel":
        return [
            {"name": "WRITE_KIND", "value": "channel"},
            {"name": "WRITE_CONVERSATION_ID", "value": conversation_id},
            {"name": "WRITE_TEAM_ID", "value": scope["teamId"]},
            {"name": "WRITE_CHANNEL_ID", "value": scope["channelId"]},
            {"name": "WRITE_THREAD_ID", "value": scope["threadId"]},
        ]
    raise ValueError(f"unknown write scope {scope!r}")


def stdio_spec(*, role: McpRole, env: list[AcpEnvVar]) -> AcpMcpServerStdio:
    command, args = mcp_server_entry()
    return {
        "type": "stdio",
        "name": server_name_for_role(role),
        "command": command,
        "args": [*args, f"--role={role}"],
        "env": env,
    }


def teams_mcp_servers(
    *,
    turn_id: TurnId,
    turns_dir: str,
    conversation_id: str,
    service_url: str,
    conversation_type: str,
    write_scope: dict[str, str],
    callback_url: str | None = None,
) -> dict[str, AcpMcpServerStdio]:
    shared = common_env(
        turn_id=turn_id,
        turns_dir=turns_dir,
        conversation_id=conversation_id,
        service_url=service_url,
        conversation_type=conversation_type,
        callback_url=callback_url,
    )
    read = stdio_spec(role="read", env=shared)
    write = stdio_spec(role="write", env=[*shared, *write_scope_env(write_scope, conversation_id)])
    return {"read": read, "write": write}


def grok_config_for_mcp(read: AcpMcpServerStdio, write: AcpMcpServerStdio) -> str:
    import json

    return json.dumps(
        {
            "mcp_servers": {
                MCP_TEAMS_READ_SERVER: {"command": read["command"], "args": read["args"]},
                MCP_TEAMS_WRITE_SERVER: {"command": write["command"], "args": write["args"]},
            }
        }
    )
